File: scala_study/src/main/scala/leetcode/minRemoveToMakeValid.scala
package leetcode


object minRemoveToMakeValid {
  def minRemoveToMakeValid(s: String): String ={
    val c = '('
    val c1 = ')'
    val length = s.length
    length
    var ans: String = ""
    var ans1: String = ""
    var balance: Int = 0
    var `sum(` = 0
    for (i <- 0 until length){

      val c2 = s.charAt(i)

      if (c2 == c){
        balance += 1
        `sum(` += 1
      }  else if (c2 == c1){
          balance-=1
      }

      if (balance < 0) {
        balance+=1
      }else{
        ans+=c2
      }

    }
    `sum(` -= balance
    for (i <- 0 until ans.length ){
      val temp=ans.charAt(i)
      if (temp == c ) `sum(`-=1

      if (`sum(` < 0){
        `sum(`+=1
      }else{
        ans1+=temp
      }

    }
    ans1
  }

  def main(args: Array[String]): Unit = {
//    val s = "lee(t(c)o)(de)(("
    val s= args(0)
    println(minRemoveToMakeValid(s))
  }
}

File: JavaLearn/src/com/leetcode/MinRemoveToMakeValid.java
package com.leetcode;

public class MinRemoveToMakeValid {

    public String minRemoveToMakeValid(String s) {
        StringBuffer stringBuffer = new StringBuffer();

        int length = s.length();
        
        int balance = 0;

        //清除不能闭合的右括号
        for (int i = 0; i < length; i++) {
            char c = s.charAt(i);
            if ("(".equals(c)){
                balance++;
            }else if (")".equals(c)){
                if (balance == 0){
                    continue;
                }else {
                    balance--;
                }
            }

            stringBuffer.append(c);
        }
        
        //清除不能闭合左括号
        String s1 = stringBuffer.reverse().toString();


        StringBuffer ans = new StringBuffer();

        int length1 = s1.length();
        balance = 0;

        for (int i = 0; i < length1; i++) {
            char c = s.charAt(i);
            if (")".equals(c)){
                balance++;
            }else if ("(".equals(c)){
                if (balance == 0){
                    continue;
                }else {
                    balance--;
                }
            }

            ans.append(c);
        }

        return ans.reverse().toString();

    }
}
